devito/types/basic.py

```diff
@@ -487,76 +487,60 @@ class AbstractTensor(sympy.ImmutableDenseMatrix, Basic, Cached, Pickable, Evalua
     is_VectorValued = False
 
     @classmethod
-    def _cache_key(cls, *args, **kwargs):
-        return cls
-
-    def __new__(cls, *args, **kwargs):
-        options = kwargs.get('options', {})
-
-        key = cls._cache_key(*args, **kwargs)
-        obj = cls._cache_get(key)
-
-        if obj is not None:
-            newobj = sympy.Matrix.__new__(cls, *args, **options)
-            newobj.__init_cached__(key)
-            return newobj
-
-        name = kwargs.get('name')
-        indices, _ = cls.__indices_setup__(**kwargs)
-
-        # Create new, unique type instance from cls and the symbol name
-        newcls = type(name, (cls,), dict(cls.__dict__))
-
-        # Create the new Function object and invoke __init__
-        comps = cls.__subfunc_setup__(*args, **kwargs)
-        newobj = sympy.ImmutableDenseMatrix.__new__(newcls, comps)
-        # Initialization. The following attributes must be available
-        newobj._indices = indices
-        newobj._name = name
-        newobj._dtype = cls.__dtype_setup__(**kwargs)
-        newobj.__init_finalize__(*args, **kwargs)
-
-        # Store new instance in symbol cache
-        Cached.__init__(newobj, newcls)
+    def _new(cls, *args, **kwargs):
+        if args:
+            try:
+                # Constructor if input is (rows, cols, lambda)
+                newobj = super(AbstractTensor, cls)._new(*args)
+            except ValueError:
+                # Constructor if input is list of list as (row, cols, list_of_list)
+                # doesn't work as it expects a flattened.
+                newobj = super(AbstractTensor, cls)._new(args[2])
```

Comment:
check `args[2]` really is what you expect it to be? or at least `if len(args) != 3 raise ValueError ...` ?

Reply:
sympy.Matrix already has proper checks and error message if it's not what expected, no reason to duplicate it.